### PR TITLE
Fe docs ecma

### PR DESCRIPTION
**Pull request by owencjones**
# Motivation and Context
Documentation for newer ECMA Script features

# What has changed
Added docs

# How to test?
Non functional, just read and comment if nesc.

# Links
https://trello.com/c/lDUOrg55

### CODE REVIEW

File: docs/frontend/Modern-JS-features.md [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+myDog.greet();
+```
+
+You can see that the above has done the following steps:
+* Replaced the class with a function - a function called with `new` in Javascript is a constructor.
+* Replaced the greet function with a similar function that is a member of `Animal.prototype` - `prototype` is an object used to create new objects.
+* Removed the string interpolation in `greet` and replaced it with concatenated strings
+* Change the use of `const` to `var`.  Some older browsers don't support `const`.
+
+> **Side note**: The above shows why `this` in Javascript behaves differently than expected.  Although in a class, you expect `this` to refer to an instance of a class, the way that objects are created with constructors is only _emulating_ classes - the functions are actually separate functions, so the reference to `this` which in Javascript refers to the scope the function exists within is not a reference to the class.
```
Comment (mfcrocker) [inline docs/frontend/Modern-JS-features.md]: `this` might just be my failure, but I found this passage hard to understand

File: docs/frontend/Modern-JS-features.md [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+      - [Asynchronicity libraries (Like async.js)](#Asynchronicity-libraries-Like-asyncjs)
+      - [Promises](#Promises)
+      - [`async` and `await` syntactic sugar](#async-and-await-syntactic-sugar)
+  - [Modules](#Modules)
+    - [CommonJS](#CommonJS)
+    - [AMD - Asynchronous Module Definition](#AMD---Asynchronous-Module-Definition)
+    - [ECMAScript modules](#ECMAScript-modules)
+
+### Introduction
+Javscript, correctly known as ECMAScript, has advanced at a fast pace in recent years, adding features like:
```
Comment (necrophonic) [inline docs/frontend/Modern-JS-features.md]: :art: Markdown lint: Should have blank line after headings

File: docs/frontend/Modern-JS-features.md [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+    - [Basic asynchronous functions](#Basic-asynchronous-functions)
+    - [Blocking vs non-blocking functions](#Blocking-vs-non-blocking-functions)
+    - [Using `async` and `await` to write more concise code](#Using-async-and-await-to-write-more-concise-code)
+      - [Asynchronicity libraries (Like async.js)](#Asynchronicity-libraries-Like-asyncjs)
+      - [Promises](#Promises)
+      - [`async` and `await` syntactic sugar](#async-and-await-syntactic-sugar)
+  - [Modules](#Modules)
+    - [CommonJS](#CommonJS)
+    - [AMD - Asynchronous Module Definition](#AMD---Asynchronous-Module-Definition)
+    - [ECMAScript modules](#ECMAScript-modules)
```
Comment (necrophonic) [inline docs/frontend/Modern-JS-features.md]: Nice use of ToC 👏

File: docs/frontend/Modern-JS-features.md [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+
+In addition to this, strings defined using the `\`` character, have other useful properties:
+* Can be written containing quote marks easily, without escaping `\`I guess that "works"\``
+* Can be multiline, which quote delimited strings can't.
+
+#### Classes
+Classes are now available in Javascript, as they are in many languages, however there is a key difference in Javascript.  Javascript classes are not a separate entity, or type, but are syntactic sugar for creating a constructor object with a prototype for building it.  This doesn't mean too much when writing the, but understanding the `prototype` object is important to writing Javascript overall.
+
+A class in JS looks like this:
+```javascript
```
Comment (necrophonic) [inline docs/frontend/Modern-JS-features.md]: :art: Should be blank line before the code block

File: docs/frontend/Modern-JS-features.md [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+
+> NB: It's important to note, the example we've used is completely predictable, but usually when using asynchronous functions, you are not using timers, but waiting on the actions of unpredictable systems
+
+##### Using `async` and `await` to write more concise code
+
+Because the asynchornicity of code is a _useful_ feature of Javascript, but the code created is not easy to read, many approaches have been used to make it simpler.  Below are some examples:
+
+###### Asynchronicity libraries (Like async.js)
+This runs `functionOne`, `functionTwo`, `functionThree` in order, despite their asynchronicity.
+```javascript
```
Comment (necrophonic) [inline docs/frontend/Modern-JS-features.md]: :art: Blank line before code block

File: docs/frontend/Modern-JS-features.md (hunk 1 above)
Comment (owencjones) [reply docs/frontend/Modern-JS-features.md]: I suppose basically classes don't exist, but code for them does - if that helps?  When you have a class, it's actually sugar for a function.

File: docs/frontend/Modern-JS-features.md (hunk 3 above)
Comment (owencjones) [reply docs/frontend/Modern-JS-features.md]: Always.  I have a plugin for generating them.

**Pull request merged**